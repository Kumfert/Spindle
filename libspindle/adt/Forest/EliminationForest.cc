//
// EliminationForest.cc
//
//  $Id: EliminationForest.cc,v 1.2 2000/02/18 01:31:44 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#include "spindle/EliminationForest.h"

#ifndef SPINDLE_PERMUTATION_H_
#include "spindle/PermutationMap.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

SPINDLE_IMPLEMENT_PERSISTANT( EliminationForest, GenericForest )

EliminationForest::EliminationForest() {
  g = 0;
  currentState = EMPTY;
}

EliminationForest::EliminationForest( const Graph* graph ) {
  incrementInstanceCount( EliminationForest::MetaData );
  g = graph;
  // set a null parent and ancestor
  const int n = g->size();
  SharedArray<int> ancestor( n );
  ancestor.init(-1);
  parent.resize(0);
  GenericForest::resize(n);
  parent.resize(n);
  parent.init(-1);
  
  for( int i=0; i<n; ++i) {
    for( const int* jj=g->begin_adj(i), *stop_jj = g->end_adj(i); 
	 jj < stop_jj; ++jj ) {
      int j = *jj;
      if ( j < i ) { // ii is an ancestor of jj
	// find the root of the tree containing the node `jj'
	while( (ancestor[j]!=-1) & (ancestor[j]!=i) ) {
	  int temp = ancestor[j];
	  ancestor[j] = i; 
	  j = temp;
	}
	// make node `ii' the root of this subtree, if not already true
	if ( ancestor[j] == -1 ) {
	  ancestor[j] = i;
	  parent[j] = i;
	}
      } // end if `ii' is ancestor of `jj'
    } // end for all jj adj to ii
  } // end for all ii
  currentState = UNKNOWN;
  createChildSibling();
  createDoubleLink();
}


EliminationForest::EliminationForest( const Graph* graph, const SharedPtr<PermutationMap>& perm ) {
  incrementInstanceCount( EliminationForest::MetaData );
  g = graph;
  
  // set a null parent and ancestor
  const int n = g->size();
  SharedArray<int> ancestor( n );
  ancestor.init(-1);
  GenericForest::resize(n);
  parent.resize(n);
  parent.init(-1);
  
  const int* old2new = perm->getOld2New().lend();
  const int* new2old = perm->getNew2Old().lend();
  
  // programmer's note:  `ii' & `jj' refer to the vertices in the new ordering
  //                      `i' & `j' refer to the same vertices in the old ordering
  
  for( int ii=0; ii<n; ++ii) {
    int i = new2old[ii];
    for( const int* jp=g->begin_adj(i), *stop_jp = g->end_adj(i); 
	 jp < stop_jp; ++jp ) {
      int j = *jp;
      int jj = old2new[ j ];
      if ( jj < ii ) { // ii is an ancestor of jj
	/* old code.  put the tree in terms of new indices, not old
	// find the root of the tree containing the node `jj'
	while( (ancestor[jj]!=-1) & (ancestor[jj]!=ii) ) {
	  int temp = ancestor[jj];
	  ancestor[jj] = ii; 
	  jj = temp;
	}
	// make node `ii' the root of this subtree, if not already true
	if ( ancestor[jj] == -1 ) {
	  ancestor[jj] = ii;
	  parent[jj] = ii;
	}
	*/
	// find the root of the tree containing the node `j'
	while( (ancestor[j]!=-1) & (ancestor[j]!=i) ) {
	  int temp = ancestor[j];
	  ancestor[j] = i; 
	  j = temp;
	}
	// make node `i' the root of this subtree, if not already true
	if ( ancestor[j] == -1 ) {
	  ancestor[j] = i;
	  parent[j] = i;
	}
      } // end if `ii' is ancestor of `jj'
    } // end for all jj adj to ii
  } // end for all ii
  currentState = UNKNOWN;
  createChildSibling();
  createDoubleLink();
}// end


EliminationForest::EliminationForest( const Graph* graph, const int n, const int* Parent )
  : GenericForest( n, Parent ) {
    incrementInstanceCount( EliminationForest::MetaData );
    g = graph;
}

EliminationForest::~EliminationForest() {
  decrementInstanceCount( EliminationForest::MetaData );
}

#define __FUNC__ "void EliminationForest::loadObject( SpindleArchive & ar )"
void
EliminationForest::loadObject( SpindleArchive & ar ) {
  ERROR( SPINDLE_ERROR_UNIMP_FUNC , "Method not yet implemented.");
  if ( &ar );
}
#undef __FUNC__

#define __FUNC__ "void EliminationForest::loadObject( SpindleArchive & ar )"
void
EliminationForest::storeObject( SpindleArchive& ar ) const {
  ERROR( SPINDLE_ERROR_UNIMP_FUNC , "Method not yet implemented.");
  if( &ar ) ;
}
#undef __FUNC__

  
