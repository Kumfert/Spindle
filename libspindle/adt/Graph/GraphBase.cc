//
//
//

#include "spindle/GraphBase.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

bool 
GraphBase::resize( const int newNVtxs, const int newNEdges, const int newNBoundVtxs=0, const int newNCutEdges=0 ) {
  
  adjHead.resize( newNVtxs + newNBoundVtxs + 1 );
  adjSelf.resize( newNVtxs + newNBoundVtxs );
  adjList.resize( newNEdges + newNCutEdges );
  if ( vtxWeight.notNull() ) { 
    vtxWeight.resize( newNVtxs );
  }
  if ( edgeWeight.notNull() ) { 
    edgeWeight.resize( 2 * newNEdges );
  }
  
  nVtxs = newNVtxs;
  nEdges = newNEdges;
  nBoundVtxs = newNBoundVtxs;
  nNonZeros = newNCutEdges + 2*newNEdges;

  return true;
}

int
GraphBase::deg( const int i ) const {
  if ( ( i>=0 ) && (i<nVtxs) ) {
    if ( vtxWeight.size() == 0 ) {
      return ( adjHead[i+1] - adjHead[i] );
    } else { 
      int temp = 0;
      for (int j = adjHead[i]; j<adjHead[i+1]; ++j) {
	if ( adjList[j] != i ) {
	  temp += vtxWeight[ adjList[ j ] ];
	}
      }
      return temp;
    }
  } else {
    return -1;
  }
}
