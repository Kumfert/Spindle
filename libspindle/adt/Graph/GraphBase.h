//
// GraphBase.h
//
//
//  Graphs are structurally symmetric, but not necessarily numerically
//  symmetric.  In the latter case, edge weights are neccessary if for
//  nothing else than to hold ones and zeros
//
//  Furthermore, graphs can be weighted
//
//  Furthermore, graphs can be partitioned.  
//  A partitioned graph stores the list of cut edges and vertices.
//
//  Not all graph algorithms are guaranteed to recognize a graph with boundary
//  vertices
//
//            nVtxs nBndryVtxs
//           +-----+---+
//           |     |   |
//           |     |   |
//           |     |   |
//           +-----+---+
//
#ifndef SPINDLE_GRAPH_BASE_H_
#define SPINDLE_GRAPH_BASE_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

SPINDLE_BEGIN_NAMESPACE
//class Graph;
class GraphBase {
  // friend class Graph;
protected:
  int nVtxs;      // number of vertices in graph... not including boundary vtxs
  int nEdges;     // number of edges in the graph ... not including cut edges
  int nBoundVtxs; // number of boundary vertices
  int nNonZeros;

  SharedArray<int> adjHead;   // length = nVtxs + nBoundVtxs + 1;
  SharedArray<int> adjSelf;   // length = nVtxs + nBoundVtxs;
  SharedArray<int> adjList;   // length = 2 (nEdges + nCutEdges) 

  SharedArray<int> vtxWeight; // length = nVtxs or zero
  SharedArray<int> edgeWeight;// length = 2 * nEdges or zero
  
  GraphBase() { } // protected default constructor

public:

  bool resize( const int newNVtxs, const int newNEdges, const int newNBoundVtxs=0, const int newNCutEdges=0 );

  int queryNVtxs() const { return nVtxs; }
  int queryNEdges() const { return nEdges; }
  int queryNNonZeros() const { return 2*nEdges; }
  int queryNBoundVtxs() const { return nBoundVtxs; }
  int queryNCutEdges() const { return adjList.size() - 2*nEdges; }

  int size() const { return  nVtxs; }
  int deg(const int i) const;
 
  //data structure access
  const SharedArray<int>& getAdjHead() const { return adjHead; }
  const SharedArray<int>& getAdjList() const { return adjList; }
  const SharedArray<int>& getAdjSelf() const { return adjSelf; } 
  const SharedArray<int>& getVtxWeight() const { return vtxWeight; }
  const SharedArray<int>& getEdgeWeight() const { return edgeWeight; }
};

SPINDLE_END_NAMESPACE

#endif
