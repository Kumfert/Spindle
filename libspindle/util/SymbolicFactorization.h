//
// SymbolicFactorization.h
//
// $Id: SymbolicFactorization.h,v 1.2 2000/02/18 01:32:04 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_SYMBOLIC_FACTORIZATION_H_
#define SPINDLE_SYMBOLIC_FACTORIZATION_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_AUTO_PTR_H_
#include "spindle/SharedPtr.h"
#endif

#ifndef SPINDLE_ALGORITHM_H_
#include "spindle/SpindleAlgorithm.h"
#endif 

#ifndef SPINDLE_PERMUTATION_H_
#include "spindle/PermutationMap.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class Graph;
class EliminationForest;

class SymbolicFactorization : public SpindleAlgorithm {

  SPINDLE_DECLARE_DYNAMIC( SymbolicFactorization )

public:
  typedef EliminationForest ETree;
  
private:
  const Graph *graph;
  SharedPtr<ETree> etree;
  SharedPtr<PermutationMap> perm;
  SharedArray<int> rowCount;
  SharedArray<int> colCount;

  bool run();
public:
  SymbolicFactorization();
  virtual ~SymbolicFactorization();

  virtual bool reset();
  bool setGraph( const Graph* g );  // sets to ready

  bool setETree( const ETree* et );  // must be ready
  bool setPermutationMap(  const PermutationMap* p ); // must be ready

  // to compute the symbolic factorization
  virtual bool execute(); // may create new permutation and etree as needed

  // to get output

  // NOTE: Even when factoring an unordered matrix, a permutation
  //       may be generated.  This permutation is a postordering of 
  //       the elimination tree.  The indices of the elimination tree
  //       are also in terms of this postordering.
  const SharedPtr<PermutationMap>& getPermutationMap() const { return perm; }
  SharedPtr<PermutationMap>& getPermutationMap() { return perm; }

  // NOTE: etree returned may not be identical to the one 
  //       provided before execution.  The factorization will 
  //       convert its copy to indices in the postordered ordering
  //       not neccessarily the original ordering
  const SharedPtr<ETree>& getETree() const { return etree; }
  SharedPtr<ETree>& getETree() { return etree; }

  long queryFill(); // = sum_{i=0}^{n} colCount[i] NNZ(A)
  long querySize(); // = sum_{i=0}^{n} colCount[i]
  double queryWork(); // = sum_{i=0}^{n} colCount[i]^2
  
};

SPINDLE_END_NAMESPACE

#endif // ifndef SPINDLE_SYMBOLIC_FACTORIZATION_H_
