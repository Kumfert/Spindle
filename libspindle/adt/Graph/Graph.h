//
// Graph.h
//
// $Id: Graph.h,v 1.2 2000/02/18 01:31:45 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997-1999, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_GRAPH_H_
#define SPINDLE_GRAPH_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_GRAPH_BASE_H_
#include "spindle/GraphBase.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

#ifdef MATLAB_MEX_FILE
extern "C" { 
#include "mex.h"
} // end extern "C"
#endif

SPINDLE_BEGIN_NAMESPACE

class MatrixBase;

class Graph: public GraphBase, public SpindlePersistant { 

  SPINDLE_DECLARE_PERSISTANT( Graph )
  
public:
  typedef const int* const_iterator;

  //constructors/destructors
  Graph();
  Graph( const int NVtxs, int *AdjHead, int *AdjList ); //assume ownership of arrays, unweighted
  Graph( const int NVtxs, const int *AdjHead, const int *AdjList ); // do not own arrays, unweighted
  Graph( const int NVtxs, int *AdjHead, int *AdjList, int *VtxWeights, int* EdgeWeights );
  Graph( const int NVtxs, const int *AdjHead, const int *AdjList, const int *VtxWeights, const int* EdgeWeights );
  Graph( GraphBase* inputGraph );
  Graph( const GraphBase* inputGraph );
  Graph( MatrixBase* inputMatrix );
  Graph( const MatrixBase* inputMatrix );
#ifdef MATLAB_MEX_FILE
  Graph( const mxArray* mat ); // constructor on matlab's Matrix
#endif 
  virtual ~Graph();

  //iterate over adjacency list of vertex i
  const_iterator begin_adj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i]) : 0 ; }
  const_iterator   end_adj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i+1]) : 0 ; }
  
  //iterate over low adjacency list of vertex i
  const_iterator begin_ladj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i]) : 0 ; }
  const_iterator   end_ladj( int i ) const { return  (isValid()) ? (adjList.begin() + adjSelf[i]) : 0 ; }

  //iterate over high adjacency list of vertex i
  const_iterator begin_hadj( int i ) const { return  (isValid()) ? (adjList.begin() + adjSelf[i]) : 0 ; }
  const_iterator   end_hadj( int i ) const { return  (isValid()) ? (adjList.begin() + adjHead[i+1]) : 0 ; }

  // required for setting variables
  virtual void validate();
  virtual void dump( FILE * fp ) const ;
  virtual bool reset();
  void prettyPrint( FILE * fp ) const ;
  //  bool isValid() const { return isValid(); } defined by SpindlePersistant

  const SharedArray<int>& getAdjHead()  const{ return adjHead; }
  const SharedArray<int>& getAdjList()  const{ return adjList; }
  const SharedArray<int>& getAdjSelf()  const{ return adjSelf; } 
  const SharedArray<int>& getVtxWeight()  const{ return vtxWeight; }
  const SharedArray<int>& getEdgeWeight()  const{ return edgeWeight; }

  SharedArray<int>& getAdjHead() { currentState=UNKNOWN; return adjHead; }
  SharedArray<int>& getAdjList() { currentState=UNKNOWN; return adjList; }
  SharedArray<int>& getAdjSelf() { currentState=UNKNOWN; return adjSelf; } 
  SharedArray<int>& getVtxWeight() { currentState=UNKNOWN; return vtxWeight; }
  SharedArray<int>& getEdgeWeight() { currentState=UNKNOWN; return edgeWeight; }

  bool setGhostNodes( const int nBoundVtxs, const int nCutEdges );
};

SPINDLE_END_NAMESPACE

#endif
