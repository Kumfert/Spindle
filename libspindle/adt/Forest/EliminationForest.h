//
// EliminationForest.h
//
//  $Id: EliminationForest.h,v 1.2 2000/02/18 01:31:44 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#ifndef SPINDLE_ARRAY_E_FOREST_H
#define SPINDLE_ARRAY_E_FOREST_H_

#ifndef SPINDLE_AUTO_PTR_H_
#include "spindle/SharedPtr.h"
#endif

#ifndef SPINDLE_ARRAY_FOREST_H_
#include "spindle/GenericForest.h"
#endif 

#ifndef SPINDLE_GRAPH_H_
#include "spindle/Graph.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class PermutationMap;

class EliminationForest : protected GenericForest { 

  SPINDLE_DECLARE_PERSISTANT( EliminationForest )

private:
  friend class constGenericForestIterator;
  friend class constGenericForestPostorderIterator;
  friend class constGenericForestPreorderIterator;

public:
  // Constructors/Destructors
  EliminationForest();
  EliminationForest( const Graph* graph );
  EliminationForest( const Graph* graph, const SharedPtr<PermutationMap>& perm );
  EliminationForest( const Graph* graph, const int n, const int* Parent );
  virtual ~EliminationForest();

private:
  const Graph* g;

public:
  typedef int Key;
  typedef int node;
  typedef constGenericForestIterator const_generic_iterator;
  typedef constGenericForestPostorderIterator const_postorder_iterator;
  typedef constGenericForestPreorderIterator const_preorder_iterator;

  virtual void validate() { GenericForest::validate(); }
  bool isValid() const { return GenericForest::isValid(); }
  void resize(const int n) { g = 0; GenericForest::resize(n); }
  bool reset() { g = 0; return GenericForest::reset(); }
  int size() const { return GenericForest::size(); }
  int queryNTrees() const { return GenericForest::queryNTrees(); }
  int queryParent( const node Child ) const { return GenericForest::queryParent(Child); }
  const SharedArray<int>& getParent() const { return GenericForest::getParent(); }

  // Iterator Access
  const_postorder_iterator begin_postorder() const { return GenericForest::begin_postorder() ; }
  const_postorder_iterator end_postorder() const { return GenericForest::end_postorder(); }
  const_preorder_iterator begin_preorder() const { return GenericForest::begin_preorder(); }
  const_preorder_iterator end_preorder() const { return GenericForest::end_preorder(); }
  const_generic_iterator create_generic_iterator() const { return GenericForest::create_generic_iterator(); }

};

SPINDLE_END_NAMESPACE 

#endif
