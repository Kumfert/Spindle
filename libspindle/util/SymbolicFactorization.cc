//
// SymbolicFactorization.h
//
// $Id: SymbolicFactorization.cc,v 1.2 2000/02/18 01:32:03 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#include "spindle/SymbolicFactorization.h"

#ifndef SPINDLE_GRAPH_H_
#include "spindle/Graph.h"
#endif

#ifndef SPINDLE_ARRAY_E_FOREST_H_
#include "spindle/EliminationForest.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

static char invalidState[] = "Invalid algorithmicState detected";
static char invalidGraph[] = "Invalid graph detected";
static char invalidETree[] = "Invalid etree detected";
static char invalidPerm[]  = "Invalid permutation detected";
static char runtimeErr[]   = "Unspecified execution error detected";
static char invalidPermFromEtree[] = "execution error detected: problem creating postordering of etree";

SymbolicFactorization::SymbolicFactorization() {
  incrementInstanceCount( SymbolicFactorization::MetaData );
  reset();
}


SymbolicFactorization::~SymbolicFactorization() {
  decrementInstanceCount( SymbolicFactorization::MetaData );
}

bool
SymbolicFactorization::reset() {
  graph = 0;
  algorithmicState= EMPTY;
  return true;
}

bool
SymbolicFactorization::setGraph( const Graph* g ) {
  if ( algorithmicState==INVALID ) { 
    return false; 
  }
  if ( ( g==0 ) || (! g->isValid()) ) { 
    algorithmicState = INVALID;
    errMsg = invalidGraph;
    return false; 
  }

  switch( algorithmicState ) {
  case DONE:
  case READY:
    reset(); // resets perm and/or etree
    // no break;
  case EMPTY:
    graph = g;
    algorithmicState = READY;
    break;
  default:
    algorithmicState = INVALID;
    errMsg = invalidState;
    return false;
  }
  return true;
}

bool
SymbolicFactorization::setETree( const ETree* et ) {
  if ( algorithmicState == INVALID ) { 
    return false; 
  }
  if ( (et == 0) || (!et->isValid()) ) { 
    algorithmicState = INVALID;
    errMsg = invalidETree;
    return false;
  }
  if ( algorithmicState == READY ) {
    if ( et->size() == graph->size() ) {
      etree.borrow( et ); // requires only const access 
      if ( perm.notNull() ) {
	perm->reset();
      }
      return true;
    }
  }
  return false;
}



bool
SymbolicFactorization::setPermutationMap( const PermutationMap* p ) {
  if ( algorithmicState == INVALID ) { 
    return false; 
  }
  if ( (p == 0) || (! p->isValid()) ) { 
    algorithmicState = INVALID;
    errMsg = invalidPerm;
    return false;
  }
  if ( algorithmicState == READY ) {
    if ( p->size() == graph->size() ) {
      perm.borrow( p ); // require only const-access
      if ( etree.notNull() ) {
	etree->reset();
      }
      return true;
    }
  }
  return false;
}


bool
SymbolicFactorization::execute() {
  if ( algorithmicState == INVALID ) { return false; }
  if ( algorithmicState != READY ) { return false; }
  
  // First we need an etree.
  if ( perm.isNull()  && etree.isNull() ) {  // if need etree and perm
    // create an etree based on the implicit ordering of the matrix
    etree.take( new ETree( graph ) );
    etree->validate();
  } else if ( perm.notNull() ) { 
    // create an elimination tree based on the permutation
    etree.take( new ETree( graph, perm ) );
    etree->validate();
  }

  if ( !etree.lend()->isValid() ) {
    algorithmicState = INVALID;
    errMsg = invalidETree;
    return false;
  }
  // ASSERT: now have a valid etree.

  // Now that we have an etree, construct a valid Permutation
  // that is a postordering of the etree
  perm.take( new PermutationMap( graph->size() ) );
  const SharedPtr<ETree>& const_etree = etree; // HACK to force const-access
  //  int * dest = perm->getNew2Old().begin();
  //  for( ETree::const_postorder_iterator it = const_etree->begin_postorder(), 
  //	 stop_it = const_etree->end_postorder(); it != stop_it; ++it ) { 
  //    *dest++ = *it;
  //    cout << *it << endl;
  //  }
  copy( const_etree->begin_postorder(), const_etree->end_postorder(), 
	perm->getNew2Old().begin() ); 
  perm->validate();
  if ( ! perm->isValid() ) {
    algorithmicState = INVALID;
    errMsg = invalidPermFromEtree;
    return false;
  }   
  // ASSERT: we now have a valid permutation and etree
  
  bool okay = run ();
  if ( !okay ) {
    algorithmicState = INVALID;
    if ( errMsg == 0 ) {
      errMsg = runtimeErr;
    }
  }
  return okay;
}

bool
SymbolicFactorization::run() {
  // PRECONDITION:  graph, permutation, and etree are all valid
  // POSTCONDITION: colCount and rowCount contain # nonzeros in factor
  //                graph, permutation and etree are unchanged

  // HACK to force const access.
  const SharedPtr<PermutationMap>& const_perm = perm;
  const SharedPtr<ETree>& const_etree = etree;

  const int n = graph->size();

  // get const access to perm, invp, and parent pointer
  register const int* old2new = const_perm->getOld2New().lend();
  register const int* new2old = const_perm->getNew2Old().lend();
  register const int* parent = const_etree->getParent().lend();

  // resize rowCount and colCount, initialize them to all 1's
  rowCount.resize(n);
  rowCount.init(1);
  colCount.resize(n);
  colCount.init(1);

  register int * row_count = rowCount.begin();
  register int * col_count = colCount.begin();

  // get a temporary array for the vertex coloring
  SharedArray<int> color(n);
  color.init(-1);
  register int * color_array = color.begin();

  // programmer's note: `ii' & `jj' refers to vertices in the new ordering
  //                     `i' &  `j' refers to the same vertices in the old ordering.
  for( int ii=0; ii<n; ++ii ) {
    color_array[ii] = ii;
    
    // get the row index in the original matrix
    int i = new2old[ii];

    // for each element in the corresponding adjacency list
    for(const int *jp = graph->begin_adj(i), *stop_jp = graph->end_adj(i);
	 jp < stop_jp; ++jp ) {
      int j = *jp;
      int jj = old2new[ j ];
      
      if ( jj < ii ) { // ii is an ancestor of jj in the permutation
	while ( color_array[jj] < ii ) { 
	  row_count[ii]++;
	  col_count[jj]++;
	  color_array[jj] = ii;
	  jj = old2new[ parent[ j ] ]; // this is right
	  if ( jj == -1 ) { // if got to root.
	    break;
	  }
	  j = new2old[ jj ];
	} // end while
      } // end if
    } // end forall j adj to i
  } // end for all ii in new ordering
  algorithmicState = DONE;
  return true;
}

long
SymbolicFactorization::querySize() {
  const int n = colCount.size();
  if ( algorithmicState == DONE ) {
    long temp = 0;
    const int *cur = colCount.begin();
    for( int i=0; i<n; ++i ) {
      temp += *cur++;
    }
    return temp;
  } else {
    return false; 
  }
}

long
SymbolicFactorization::queryFill() {
  const int n = colCount.size();
  if ( algorithmicState == DONE ) {
    long temp = 0;
    const int *cur = colCount.begin();
    for( int i=0; i<n; ++i ) {
      temp += *cur++;
    }
    temp -=  graph->queryNEdges() + graph->queryNVtxs();
    return temp;
  } else {
    return false; 
  }
}

double
SymbolicFactorization::queryWork() {
  const int n = colCount.size();
  if ( algorithmicState == DONE ) {
    double temp = 0.0;
    for( int i=0; i<n; ++i ) {
      double j = colCount[i];
      temp += (j-1.0) * (j+2.0);
    }
    return temp;
  } else {
    return false; 
  }
}


SPINDLE_IMPLEMENT_DYNAMIC( SymbolicFactorization, SpindleAlgorithm )

