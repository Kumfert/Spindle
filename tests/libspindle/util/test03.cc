//
// testSymbolicFactorization.cc
//
//
//

#include <iostream.h>
#include "spindle/Graph.h"
#include "spindle/EliminationForest.h"
#include "spindle/SymbolicFactorization.h"

#include "samples/small_mesh.h"

static const int mmd_new2old_perm[]   = { 0, 2, 6, 8, 1, 7, 3, 4, 5 };
static const int mmd_parent_pointer[] = { 1, 3, 1, 4, 5, 5, 7, 3, 7 }; 
#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

int main() {
  ios::sync_with_stdio();
  typedef SymbolicFactorization::ETree ETree;

  Graph graph(nvtxs,Aptr,Aind);
  graph.validate();

  if ( ! graph.isValid() ) {
    cerr << "Cannot create valid graph." << endl;
  }

  // first try for an unordered graph
  {
    SymbolicFactorization symbfact;
    
    if ( ! symbfact.setGraph( &graph ) ) {
      cerr << "Error setting Graph" << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    if ( ! symbfact.execute() ) {
      cerr << "Error doing symbolic factorization on unordered graph." << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    
    cout << "For unordered graph: " << endl;
    cout << "     Size of the factor     = " << symbfact.querySize() << endl;
    cout << "     Size of fill           = " << symbfact.queryFill() << endl;
    cout << "     Amount of work (flops) = " << symbfact.queryWork() << endl;
    cout << endl;
  }

  // now try again with the graph permuted
  {
    SymbolicFactorization symbfact;
    
    if ( ! symbfact.setGraph( &graph ) ) {
      cerr << "Error setting Graph" << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    PermutationMap* p = new PermutationMap(9);
    p->getNew2Old().import( mmd_new2old_perm , 9 );
    p->validate();
    if (! p->isValid() ) {
      cerr << "PermutationMap Object creation failed." << endl;
      exit(-1);
    }
    if ( ! symbfact.setPermutationMap( p ) ) {
      cerr << "Error seting permutation in symbolic factorization" << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    if ( ! symbfact.execute() ) {
      cerr << "Error doing symbolic factorization on unordered graph." << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    
    cout << "For permuted graph: " << endl;
    cout << "     Size of the factor     = " << symbfact.querySize() << endl;
    cout << "     Size of fill           = " << symbfact.queryFill() << endl;
    cout << "     Amount of work (flops) = " << symbfact.queryWork() << endl;
    cout << endl;
    delete p;
  }

  // finally with a predefined elimination tree
  {
    SymbolicFactorization symbfact;
    
    if ( ! symbfact.setGraph( &graph ) ) {
      cerr << "Error setting Graph" << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    ETree* et = new ETree( &graph, graph.size(), mmd_parent_pointer );
    et->validate();
    if (! et->isValid() ) {
      cerr << "Elimination Forest Object creation failed." << endl;
      exit(-1);
    }
    if ( ! symbfact.setETree( et ) ) {
      cerr << "Error seting elimination tree in symbolic factorization" << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    if ( ! symbfact.execute() ) {
      cerr << "Error doing symbolic factorization on unordered graph." << endl;
      cerr << symbfact.queryErrMsg() << endl;
      exit(-1);
    }
    
    cout << "For graph with etree: " << endl;
    cout << "     Size of the factor     = " << symbfact.querySize() << endl;
    cout << "     Size of fill           = " << symbfact.queryFill() << endl;
    cout << "     Amount of work (flops) = " << symbfact.queryWork() << endl;
    cout << endl;
    delete et;
  }
}
