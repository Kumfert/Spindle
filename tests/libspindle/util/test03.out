For unordered graph: 
     Size of the factor     = 29
     Size of fill           = 8
     Amount of work (flops) = 114

For permuted graph: 
     Size of the factor     = 26
     Size of fill           = 5
     Amount of work (flops) = 90

For graph with etree: 
     Size of the factor     = 26
     Size of fill           = 5
     Amount of work (flops) = 90

